docs/langref/relay_type.rst

```diff
@@ -239,3 +239,133 @@ parameters: Type parameters must be bound to a function and are replaced with co
 at call sites, whereas incomplete types may appear anywhere in the program and are filled in during type inference.
 
 See :py:class:`~tvm.relay.ty.IncompleteType` for its definition and documentation.
+
+.. _adt-typing:
+
+Algebraic Data Types
+====================
+
+*Note: ADTs are not currently supported in the text format.*
+
+Algebraic data types (ADTs) are described in more detail in
+`their overview <adt-overview_>`__; this section describes
+their implementation in the type system.
+
+An ADT is defined by a collection of named constructors,
+each of which takes arguments of certain types.
+An instance of an ADT is a container that stores the values
+of the constructor arguments used to produce it as well as the
+name of the constructor; the values can be retrieved by
+deconstructing the instance by matching based on its constructor
+(hence, ADTs are sometimes called "tagged unions").
+
+From the type system's perspective, it is most pertinent that
+ADTs can take type parameters (constructor arguments can be
+type parameters, though ADT instances with different type
+parameters must be treated as different types) and be
+recursive (a constructor for an ADT can take an instance of
+that ADT, thus an ADT like a tree or list can be inductively
+built up). The representation of ADTs in the type system must
+be able to accomodate these facts, as the below sections will detail.
+
+Global Type Variable
+~~~~~~~~~~~~~~~~~~~~
+
+To represent ADTs compactly and easily allow for recursive ADT definitions,
+an ADT definition is given a handle in the form of a global type variable
+that uniquely identifies it. Each ADT definition is given a fresh global
+type variable as a handle, so pointer equality can be used to distinguish
+different ADT names.
+
+For the purposes of Relay's type system, ADTs are differentiated by name;
+that means that if two ADTs have different handles, they will be
+considered different types even if all their constructors are
+structurally identical.
+
+Recursion in an ADT definition thus follows just like recursion for a
+global function: the constructor can simply reference the ADT handle
+(global type variable) in its definition.
+
+See :py:class:`~tvm.relay.ty.GlobalTypeVar` for its definition and documentation.
+
+Definitions (Type Data)
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Besides a name, an ADT needs to store the constructors that are used
+to define it and any type paramters used within them. These are
+stored in the module, `analogously to global function definitions <module-description_>`__.
+
+While type-checking uses of ADTs, the type system sometimes must
```

Comment:
This is too vague. Example?

Reply:
Examples are given later, but maybe I'll move one up.

Reply:
I'll see what I can say more specifically about it